### PR TITLE
[RFC] options: remove nonfunctional &termencoding option

### PR DESCRIPTION
**Pull request by bfredl**
As discussed in #2627, remove obsolete `&termencoding` option.

### CODE REVIEW

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -6569,36 +6567,7 @@ A jump table for the options with a short description can be found at |Q_op|.
...
-	|encoding-table|.
-	The value for this option must be supported by internal conversions or
-	iconv().  When this is not possible no conversion will be done and you
-	will probably experience problems with non-ASCII characters.
-	Example: You are working with the locale set to euc-jp (Japanese) and
-	want to edit a UTF-8 file: >
-		:let &termencoding = &encoding
-		:set encoding=utf-8
-<	You need to do this when your system has no locale support for UTF-8.
+'termencoding' 'tenc'	Removed. {Nvim}
```
Comment (Pyrohh) [inline runtime/doc/options.txt]: Should be mentioned in `vim_diff.txt`, and have a `|vim-differences|` tag (see the other instances in `options.txt`).

**Comment by Pyrohh** [comment]
There are many references to this in the docs (`git grep -w tenc` should help) as well as syntax definitions.

**Comment by bfredl** [comment]
A lot of `mbyte.txt` is not up to date (much about terminal/gui/windows), but I will only remove the`termencoding` refererences in this PR. Cleaning up these docs ("Don't be surprised if it takes
quite a bit of work and experimenting to make Vim use all the multi-byte
features." hopefully not for most nvim setups....) will probably require more discussion in a separate PR.

**Comment by bfredl** [comment]
similarily with `usr_45.txt`, which tells you "USING UNICODE IN AN ORDINARY TERMINAL" where "ordinary" means lacking unicode support...

File: runtime/doc/usr_45.txt [hunk 2]
```diff
@@ -234,16 +222,7 @@ Unicode support.  You can do this with Vim, although characters that are not
...
-
-This is the same as what was used for the GUI.  But it works differently: Vim
-will convert the displayed text before sending it to the terminal.  That
-avoids that the display is messed up with strange characters.
-   For this to work the conversion between 'termencoding' and 'encoding' must
-be possible.  Vim will convert from latin1 to Unicode, thus that always works.
-For other conversions the |+iconv| feature is required.
-   Try editing a file with Unicode characters in it.  You will notice that Vim
+Try editing a file with Unicode characters in it.  You will notice that Vim
 will put a question mark (or underscore or some other character) in places
```
Comment (bfredl) [inline runtime/doc/usr_45.txt]: I don't think the new TUI does this?

**Comment by bfredl** [comment]
@Pyrohh updated with you suggestions. Removed all references to `termencoding`/`tenc` (except `todo.txt` which I didn't dare touch...)

File: runtime/doc/mbyte.txt [hunk 3]
```diff
@@ -100,8 +100,7 @@ See 'fileencoding', 'fileencodings' and |++enc|.
 DISPLAY AND FONTS
 
 If you are working in a terminal (emulator) you must make sure it accepts the
-same encoding as which Vim is working with.  If this is not the case, you can
-use the 'termencoding' option to make Vim convert text automatically.
+same encoding as which Vim is working with.
```
Comment (Pyrohh) [inline runtime/doc/mbyte.txt]: trailing whitespace at EOL.

File: runtime/doc/usr_45.txt [hunk 4]
```diff
@@ -234,16 +222,7 @@ Unicode support.  You can do this with Vim, although characters that are not
 supported by the terminal will not be displayed.  The layout of the text
 will be preserved.  >
```
Comment (Pyrohh) [inline runtime/doc/usr_45.txt]: the `>` can be removed.

**Comment by Pyrohh** [comment]
> A lot of mbyte.txt is not up to date (much about terminal/gui/windows), but I will only remove thetermencoding refererences in this PR

I think that's fine; I skimmed it and it is indeed very outdated.

Anyways, besides the mentioned nits this LGTM.

**Comment by Shougo** [comment]
> A lot of mbyte.txt is not up to date (much about terminal/gui/windows), but I will only remove thetermencoding refererences in this PR. Cleaning up these docs ("Don't be surprised if it takes
quite a bit of work and experimenting to make Vim use all the multi-byte
features." hopefully not for most nvim setups....) will probably require more discussion in a separate PR.

Hi.  I'm multi-byte user.
It should be careful to remove multi-byte features.
Yes, I think neovim can remove GUI multibyte support description is removable.

**Comment by Shougo** [comment]
If `&termencoding` is not effected in neovim, most of their implementation can remove.
The documentation can change `"It is obsolete option.  It is no effect in neovim."`
But neovim should not remove it.  Because, it is used in some of Vim plugins.

**Comment by bfredl** [comment]
> It should be careful to remove multi-byte features.

True, this is mostly about features that's no longer implemented in nvim core, rather by the GUI:s. For instance nvim do not need to deal with input methods directly, but nvim gui:s, and terminal emulators need to.

> But neovim should not remove it[&termencoding]. Because, it is used in some of Vim plugins.

Any examples? I can't see why though, plugins deal with `&enc`oded text, not `&termenc`oded text.

**Comment by Shougo** [comment]
> Any examples? I can't see why though, plugins deal with &encoded text, not &termencoded text.

OK. It is vim-quickrun.  neovim should not break existing Vim plugins.

https://github.com/thinca/vim-quickrun/search?utf8=%E2%9C%93&q=termencoding

**Comment by bfredl** [comment]
It seem to reencode shell commands from `&enc` to `&tenc` though I cannot understand why. Is it using `&tenc` as a substitute for "the encoding of the envirionment" ? I don't think that's correct, already in gvim `&tenc` is always utf-8 regardless, so that plugin won't work with gvim if the shell envirionment is not utf-8. (one should supply "" to iconv to mean locale encoding, IIRC)

Before other options has been removed. It could be made "hidden" of course, but I don't see the difference with earlier removed options.

**Comment by bfredl** [comment]
I was wrong, `iconv(..., &enc,  "")` is broken even in vim... (inconsistent with cmdline `iconv`). one should use `iconv(..., &enc, "default")` for this purpose.

File: runtime/doc/usr_45.txt (hunk 4 above)
Comment (bfredl) [inline runtime/doc/usr_45.txt]: what `>` ? :) oh, it was both concealed _and_ invisible... Perhaps the later (fg==bg) should not happen when _editing_ help files.

File: runtime/doc/mbyte.txt (hunk 3 above)
Comment (bfredl) [inline runtime/doc/mbyte.txt]: :+1:

**Comment by Shougo** [comment]
@bfredl I think current vim-quickrun implementation is wrong.  It should use "char" for the conversion.
But the problem is not it.
My problem is "don't remove the option because it is already used option".
If the plugins or users use/change the option, the error messages are occurred.
It breaks compatibilities.

**Comment by bfredl** [comment]
latest commit marks it "hidden" instead. Another option could be to throw an error if `&tenc` is set _differently_ to `&enc`, but I'm not sure that can be done reliably and really worth it. One could argue that it is like `&guifont`, it should be just hidden because it is not relevant anymore.

**Comment by bfredl** [comment]
Squashed and rebased. I will merge after tests if no more comments.

**Pull request closed without merging**